FileShareAdminClient/FileShareApiAdminClient.cs

```diff
@@ -245,24 +277,136 @@ public async Task RollBackBatchAsync(IBatchHandle batchHandle)
             }
             ((BatchHandle)batchHandle).AddFile(fileName, Convert.ToBase64String(md5Hash));
         }
+        private async Task<IResult<AddFileToBatchResponse>> AddFiles(IBatchHandle batchHandle, Stream stream, string fileName, string mimeType,
+            Action<(int blocksComplete, int totalBlockCount)> progressUpdate, CancellationToken cancellationToken,
+            params KeyValuePair<string, string>[] fileAttributes)
+        {
+            var mappedResult = new Result<AddFileToBatchResponse>();
+            if (!stream.CanSeek)
+                throw new ArgumentException("The stream must be seekable.", nameof(stream));
+            stream.Seek(0, SeekOrigin.Begin);
+
+            var fileUri = $"batch/{batchHandle.BatchId}/files/{fileName}";
+            {
+                var fileModel = new FileModel()
+                { Attributes = fileAttributes ?? Enumerable.Empty<KeyValuePair<string, string>>() };
+
+                Dictionary<string, string> requestHeaders = new Dictionary<string, string>();
+                requestHeaders.Add("X-Content-Size", "" + stream.Length);
+                if (!string.IsNullOrEmpty(mimeType)) requestHeaders.Add("X-MIME-Type", mimeType);
+
+                var result = await SendResult<FileModel, AddFileToBatchResponse>(fileUri, HttpMethod.Post, fileModel, cancellationToken,
+                    HttpStatusCode.Created, false, requestHeaders);
+                if (result.Errors != null && result.Errors.Any())
+                {
+                    mappedResult = (Result<AddFileToBatchResponse>)result;
+                }
+                else
+                {
+                    var md5Hash = stream.CalculateMD5();
+                    stream.Seek(0, SeekOrigin.Begin);
+
+                    var fileBlocks = new List<string>();
+                    var fileBlockId = 0;
+                    var expectedTotalBlockCount = (int)Math.Ceiling(stream.Length / (double)maxFileBlockSize);
+                    progressUpdate((0, expectedTotalBlockCount));
+
+                    var buffer = new byte[maxFileBlockSize];
+                    while (true)
+                    {
+                        fileBlockId++;
+                        var ms = new MemoryStream();
+
+                        var read = stream.Read(buffer, 0, maxFileBlockSize);
+                        if (read <= 0) break;
+                        ms.Write(buffer, 0, read);
+
+                        var fileBlockIdAsString = fileBlockId.ToString("D5");
+                        var putFileUri = $"batch/{batchHandle.BatchId}/files/{fileName}/{fileBlockIdAsString}";
+                        fileBlocks.Add(fileBlockIdAsString);
+                        ms.Seek(0, SeekOrigin.Begin);
+
+                        var blockMD5 = ms.CalculateMD5();
+
+                        using (var httpRequestMessage = new HttpRequestMessage(HttpMethod.Put, putFileUri)
```

Comment:
I think we should split the `SendResult` method into two smaller methods, one method for JSON requests and another method for sending fully populated `HttpRequestMessage`s.
Doing this will have a few benefits:
- We won't need to create a complete `HttpRequestMessage` in `AddFiles`, only to split it apart, and then re-assemble it in `SendResult`.
- The complexity in `SendResult` remains low

I have created a PR to better explain what I mean:
https://github.com/UKHO/file-share-dotnet-client/pull/23

Reply:
Code updated as per suggestion.